plasmapy/atomic/atomic.py

```diff
@@ -469,109 +365,100 @@ def ion_mass(argument: Union[str, int, Quantity], Z: int = None,
-            return m_i
-
-    if _is_electron(argument) or _is_positron(argument):
-        return const.m_e
-    elif _is_proton(argument, Z, mass_numb) or _is_antiproton(argument):
+    if particle.ion or particle.particle in {'e+'}:
+        return particle.mass
+    elif particle.particle == 'n':
+        raise InvalidIonError
+    elif particle.isotope in ['H-1']:
+        warnings.warn("Use 'p+' instead of 'H-1' to refer to protons",
+                      DeprecationWarning)
         return const.m_p
-    elif _is_neutron(argument, mass_numb):
-        raise InvalidIonError("Use isotope_mass or m_n to get mass of neutron")
-
-    if isinstance(argument, str):
-        argument, Z_from_arg = _extract_integer_charge(argument)
+    elif particle.element:
+        warnings.warn("The assumption that particles are singly ionized in"
+                      "ion_mass is deprecated.", DeprecationWarning)
+        return particle.mass - const.m_e
+    elif particle.mass is not None:
+        warnings.warn('The use of ion_mass for particles besides ions and '
+                      'positrons is deprecated.', DeprecationWarning)
+        return particle.mass
     else:
-        Z_from_arg = None
-
-    if Z is None and Z_from_arg is None:
-        Z = 1
-    elif Z is not None and Z_from_arg is not None and Z != Z_from_arg:
-        raise InvalidIonError("Inconsistent charge information in"
-                              "ion_mass.")
-    elif Z is None and Z_from_arg is not None:
-        Z = Z_from_arg
-
-    if isinstance(Z, str) and Z.isdigit():
-        Z = int(Z)
-    if isinstance(mass_numb, str) and mass_numb.isdigit():
-        mass_numb = int(mass_numb)
-
-    if not isinstance(Z, int):
-        raise TypeError("In ion_mass, Z must be an integer representing the "
-                        "ionization state (e.g., Z=1 for singly ionized).")
-
-    if not isinstance(mass_numb, int) and mass_numb is not None:
-        raise TypeError("In ion_mass, mass_numb must be an integer "
-                        "representing the mass number of an isotope.")
-
-    if atomic_number(argument) < Z:
-        raise InvalidParticleError("The ionization state cannot exceed the "
-                                   "atomic number in ion_mass")
-
-    try:
-        isotope = isotope_symbol(argument, mass_numb)
-    except InvalidParticleError as e:
-        raise InvalidParticleError("Invalid particle in ion_mass.")
-    except InvalidIsotopeError:
-        is_isotope = False
-    else:
-        is_isotope = True
+        raise InvalidIonError
 
-    if is_isotope:
 
-        if isotope == 'H-1' and Z == 1:
-            return const.m_p
-        elif isotope == 'D' and Z == 1:
-            return 3.343583719e-27 * u.kg
-        elif isotope == 'T' and Z == 1:
-            return 5.007356665e-27 * u.kg
+@particle_input(exclude={'neutrino', 'antineutrino'})
+def particle_mass(particle: Particle, *, Z: int = None,
```

Comment:
For this I was intending to not allow `Z` and `mass_numb` to be positional arguments, since they're both integers and could get confused with each other.

Reply:
oh yeah of course!